### ISSUE TITLE
Question: How is Huggingface doing language modeling?

### ISSUE THREAD

**Issue by rmorain** [rmorain/kirby#24]
Is it doing windowing? Something else? What is the loss?

**Comment by rmorain** [comment]
Here are some details I am finding:

GPT2 with the language model head outputs hidden_states which is then run through a linear layer to get logits for each word in the vocabulary. It is of shape (batch_size, sequence_length, vocab_size). The labels are of size (batch_size, sequence length). So there is a label for each word in the sequence.

These logits are then shifted (left or right?) so now they are of size (batch_size, sequence_length - 1, vocab_size). 
The labels are also shifted (batch_size, sequence_length - 1).

The shift logits are then reshaped to be (batch_size x sequence_length - 1, vocab_size)
Labels are flattened to (batch_size x sequence_length - 1)

These are then run through the cross entropy loss function.

This is my understanding for what is going on. 

The purpose of the shift is to use the n-1 words to predict the nth word. 

In short, I think everything will be fine. I did discover that our data processing may have some problems. I have created  https://github.com/rmorain/kirby/issues/32 to address this.

We can also use the Huggingface example as a baseline to ensure our GPT2 model is performing correctly.
